### PR TITLE
Change article.tries system and read multiple articles from queue

### PR DESCRIPTION
**Pull request by puzzledsab**
Issue #1006. Instead of resetting the `try_list` I have added an `article_queue` list to the server object, which the article is added to. I'm not sure how well this works because I don't know how to test the various retry scenarios properly. One issue I have seen is that if the nzb is deleted while it is being downloaded, the articles currently being downloaded will be retried.

`article_queue` can hopefully also be used to read a few articles in advance from `nzf.get_article()`, which would require a bit less resources than doing the full `NzbQueue.get_article()` every time.

### CODE REVIEW

**Comment by puzzledsab** [comment]
I've found a problem that can possibly cause problems with the current version as well. When it's paused it will put the articles being downloaded back in the `article_queue`. I thought `__reset_nw` would stop the downloading so I tried to set `fetcher = None` for the articles in article_queue. However, the downloading can still continue, causing
```
Exception in thread Thread-5:
Traceback (most recent call last):
  File "C:\Users\t\AppData\Local\Programs\Python\Python38\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\gittestbranch\sabnzbd\sabnzbd\downloader.py", line 822, in run
    self.decode(article, nw.data)
  File "C:\gittestbranch\sabnzbd\sabnzbd\downloader.py", line 447, in decode
    sabnzbd.BPSMeter.register_server_article_tried(article.fetcher.id)
AttributeError: 'NoneType' object has no attribute 'id'
```
I haven't found a way to fix this yet, so maybe this PR no closer to a solution.

Anyway, I think this may be the reason for a problem I have seen once in a while when pausing and resuming. Some files will not finish downloading until I restart SAB. I suspect it's because the nw is put back into idle_threads with an unfinished article. It doesn't finish downloading so it's not sent to the decoder and article.fetcher is not reset so it's not picked up again by `get_article`.

**Comment by puzzledsab** [comment]
Ignore the exception, I messed up in my first attempt.

File: sabnzbd/downloader.py [hunk 1]
```diff
@@ -691,7 +710,8 @@ def run(self):
                                     # Don't count this for the tries (max_art_tries) on this server
                                     self.__reset_nw(nw, count_article_try=False, send_quit=True)
                                     self.plan_server(server, _PENALTY_TOOMANY)
-                                    server.threads -= 1
+                                    if server.threads > 1:
+                                        server.threads -= 1
```
Comment (Safihre) [inline sabnzbd/downloader.py:714]: Why is this needed? Can it go below 0? That sounds like a problem somewhere else.

**Comment by Safihre** [comment]
Looking at this, I don't want to merge this just for #1006. 
However, it seems this would really be a good base to extend it to fetch `server.threads` number of articles at once.

File: sabnzbd/downloader.py (hunk 1 above)
Comment (puzzledsab) [reply sabnzbd/downloader.py:714]: It's to stop it from going below 1, not below 0. It seems strange if 1 connection is too many.

**Comment by thezoggy** [comment]
Seems like we are forcing it everywhere now? Maybe we should have just done force is default true unless explicitly passed not to

**Comment by puzzledsab** [comment]
There doesn't seem to be any point in making the prefetching more complicated than the last update. It reduces the CPU usage of NzbQueue.get_article and NzbObject.get_article to a tiny fraction. It wasn't very high to start with, though. In total this update reduces the total CPU usage by a bit more than 1% in my tests.

I have just realized that the reason why I thought the get_article functions used so much CPU was that I had configured a top priority server with 1 day retention (for testing the pickling). That made it do 33 times as many article.get_article calls as without that server if the nzb is more than 1 day old. This seems a bit excessive, but it's a more limited problem which probably doesn't affect very many users.

**Comment by Safihre** [comment]
>  In total this update reduces the total CPU usage by a bit more than 1% in my tests.

Interesting. I am surprised a bit. I always thought the `get_article` function was slow due to all the checks and called _a lot_. 
What I had in mind was kind of like what you made a long time ago: to not just pre-fetch 1 article, but maybe 5 at once. So also modify the `get_article` function.
But if you say that the time spent in those functions is negligible, it might indeed not be worth it to complicate things.

**Comment by puzzledsab** [comment]
Rewrote to fetch multiple articles from nzf.get_article(). It doesn't matter when they are fetched for the top server because then it will be the next article in line in the nzf, but it's a huge improvement when it's not. When I have a 1 day retention server in front which will be marked as tried immediately it reduces the number of calls to article.get_article from 3.4M to 200K on my test NZB.

**Comment by puzzledsab** [comment]
I realized that fetching a lot of articles if there are many connections may be counter productive so I set it to a fixed limit of 4. Under normal circumstances it will only be marginally faster to fetch 4 new articles in 1 call than to fetch 1 new article 4 times using nzf.get_articles. Fetching too many articles in one call may take enough time that the delay slows down the download speed on fast connections.

**Comment by puzzledsab** [comment]
This version needs too many changes to work with my new pickler and also has some bugs I have fixed in the other branch. https://github.com/puzzledsab/sabnzbd/tree/feature/optimized
